File: Algorithms/0150_Evaluate_Reverse_Polish_Notation/Java/EvaluateReversePolishNotationSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int evalRPN(String[] tokens) {
        if (tokens.length == 1) return Integer.parseInt(tokens[0]);
        Stack<String> stack = new Stack<>();
        HashSet<String> operators = Stream.of("+", "-", "*", "/").collect(Collectors.toCollection(HashSet::new));
        for (String token : tokens) {
            if (operators.contains(token)) {
                String right = stack.pop();
                String left = stack.pop();
                stack.add(calc(left, right, token));
            } else {
                stack.add(token);
            }
        }
        return Integer.parseInt(stack.peek());
    }

    private String calc(String left, String right, String operator) {
        int leftNum = Integer.parseInt(left);
        int rightNum = Integer.parseInt(right);
        if (operator.equals("+")) return Integer.toString(leftNum + rightNum);
        if (operator.equals("-")) return Integer.toString(leftNum - rightNum);
        if (operator.equals("*")) return Integer.toString(leftNum * rightNum);
        if (operator.equals("/")) return Integer.toString(leftNum / rightNum);
        return "";
    }
}

File: Algorithms/0150_Evaluate_Reverse_Polish_Notation/Python/Evaluate_Reverse_Polish_Notation_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def evalRPN(self, tokens):
        length = len(tokens)
        if length == 1 and tokens[0].isdecimal(): return int(tokens[0])
        stack = []
        operators = {'+', '-', '*', '/'}
        for i in tokens:
            if i in operators:
                right = stack.pop()
                left = stack.pop()
                res = str(int(eval(left + i + right)))
                stack.append(res)
            else:
                stack.append(i)
        return int(stack[0])




